polling_stations/apps/data_collection/data_types.py

```diff
@@ -164,6 +152,91 @@ def append(self, address):
                 pretty=True,
             )
 
+    def validate_postcodes(self, addressbase_data):
+        """
+        This method checks if a record has a valid postcode.
+        If it doesn't then it looks for uprn, and if it can match the uprn in
+        addressbase, and the addresses look similar it will correct the postcode
+        for us.
+        Even if it doesn't match the uprn it saves us some legwork by having tried
+        various options.
+        """
+
+        postcode_validator = GBPostcodeField()
+
+        def correct_postcode_from_uprn(record, addressbase_data):
+            addressbase_record = addressbase_data[record["uprn"]]
+            match_quality = fuzz.partial_ratio(
+                record["address"].lower().replace(",", ""),
+                addressbase_record["address"].lower().replace(",", ""),
+            )
+            if match_quality >= 100:
+                # Retry validator.clean
+                try:
+                    postcode_validator.clean(addressbase_record["postcode"])
+                    record["postcode"] = addressbase_record["postcode"]
+                    self.logger.log_message(
+                        logging.INFO,
+                        "Replacing %s with %s for record:\n%s\n",
+                        variable=(
+                            record["postcode"],
+                            addressbase_record["postcode"],
+                            record,
+                        ),
+                    )
+                    return True
+                except ValidationError:
```

Comment:
I reckon we can drop the second call to `clean()`/`except ValidationError` here and assume that if the postcode came from addressbase, it is valid. We make the same assumption everywhere else in the codebase.

Reply:
Done :)